python/ray/data/datasource/tfrecords_datasource.py

```diff
@@ -0,0 +1,78 @@
+import tempfile
+from typing import Dict, List, Union
+import warnings
+
+import os
+import pandas as pd
+import pyarrow
+import tensorflow as tf
+
+from ray.data.block import Block
+from ray.data.datasource.file_based_datasource import FileBasedDatasource
+
+
+class TFRecordDatasource(FileBasedDatasource):
+
+    _FILE_EXTENSION = "tfrecords"
+
+    def _read_file(self, f: "pyarrow.NativeFile", path: str, **reader_args) -> Block:
+        from google.protobuf.message import DecodeError
+
+        with tempfile.NamedTemporaryFile() as temporary_file:
+            # TensorFlow doesn't expose a function for reading TFRecords with a file
+            # handle; you can only read TFRecords with a file path. So, if the TFRecord
+            # doesn't exist locally (e.g., because it's stored in S3), we have to
+            # download a copy.
```

Comment:
I think that it would be better if we put in the engineering effort to manually parse the `TFRecord` file ourselves, since this download locally + read file is going to be very inefficient.

Ideally we could port TensorFlow's TFRecord reader to Python, within this datasource: https://github.com/tensorflow/tensorflow/blob/769eddaf479c8debead9a59a72617d6ed6f0fe10/tensorflow/core/lib/io/record_reader.cc#L88-L210

Here is the TFRecord file format details: https://www.tensorflow.org/tutorials/load_data/tfrecord#tfrecords_format_details

And here is an example of a Python port of the TFRecord reader: https://github.com/vahidk/tfrecord/blob/74b2d24a838081356d993ec0e147eaf59ccd4c84/tfrecord/reader.py#L16-L223

If you think this would be a lot to take on in this initial PR, we can have this as a P1 follow-up.

Reply:
Fixed